src/ApparentHorizons/StrahlkorperGr.cpp

```diff
@@ -6,15 +6,353 @@
+
+  auto deriv_surface_metric =
+      make_with_value<tnsr::ijj<DataVector, 2, Frame::Spherical<Fr>>>(
+          get<0, 0>(surface_metric), 0.0);
+  // Get the partial derivative of the metric from the Pfaffian derivative
+  get<0, 0, 0>(deriv_surface_metric) = get<0>(grad_surface_metric_theta_theta);
+  get<1, 0, 0>(deriv_surface_metric) =
+      get(sin_theta) * get<1>(grad_surface_metric_theta_theta);
+  get<0, 0, 1>(deriv_surface_metric) = get<0>(grad_surface_metric_theta_phi);
+  get<1, 0, 1>(deriv_surface_metric) =
+      get(sin_theta) * get<1>(grad_surface_metric_theta_phi);
+  get<0, 1, 1>(deriv_surface_metric) = get<0>(grad_surface_metric_phi_phi);
+  get<1, 1, 1>(deriv_surface_metric) =
+      get(sin_theta) * get<1>(grad_surface_metric_phi_phi);
+
+  const auto trace_christoffel_second_kind =
+      trace_last_indices(raise_or_lower_first_index(
+                             gr::christoffel_first_kind(deriv_surface_metric),
+                             inverse_surface_metric),
+                         inverse_surface_metric);
+
+  return trace_christoffel_second_kind;
+}
+
+// I'm going to solve a general eigenvalue problem of the form
+// A x = lambda B x, where A and B are NxN, where N is the
+// number of elements with l > 0 and l < ntheta - 2,
+// i.e. l < l_max + 1 - 2 = l_max - 1. This function computes N.
+size_t get_matrix_dimension(const YlmSpherepack& ylm) noexcept {
+  // If l_max == m_max, there are square(l_max+1) Ylms total
+  size_t matrix_dimension = square(ylm.l_max() + 1);
+  // If l_max > m_max, there are
+  // (l_max - m_max) * (l_max - m_max + 1) fewer Ylms total
+  matrix_dimension -=
+      (ylm.l_max() - ylm.m_max()) * (ylm.l_max() - ylm.m_max() + 1);
+  // The actual matrix dimension is smaller, because we do not count
+  // Ylms with l == 0, l == l_max, or l == l_max - 1.
+  matrix_dimension -= 4 * ylm.m_max() + 3;
+  if (ylm.l_max() == ylm.m_max()) {
+    matrix_dimension += 2;
+  }
+  return matrix_dimension;
+}
+
+// Get left matrix A and right matrix B for eigenproblem A x = lambda B x.
+template <typename Fr>
+void get_left_and_right_eigenproblem_matrices(
+    const gsl::not_null<Matrix*> left_matrix,
+    const gsl::not_null<Matrix*> right_matrix,
+    const tnsr::II<DataVector, 2, Frame::Spherical<Fr>>& inverse_surface_metric,
+    const tnsr::I<DataVector, 2, Frame::Spherical<Fr>>&
+        trace_christoffel_second_kind,
+    const Scalar<DataVector>& sin_theta, const Scalar<DataVector>& ricci_scalar,
+    const YlmSpherepack& ylm) noexcept {
+  const auto grad_ricci_scalar = ylm.gradient(get(ricci_scalar));
+  // loop over all terms with 0<l<l_max-1: each makes a column of
+  // the matrices for the eigenvalue problem
+  auto iter_i = SpherepackIterator(ylm.l_max(), ylm.m_max());
+  auto iter_j = SpherepackIterator(ylm.l_max(), ylm.m_max());
+  size_t column = 0;  // number which column of the matrix we are filling
+  for (iter_i.reset(); iter_i; ++iter_i) {
```

Comment:
Why can't you just define `iter_i` here? Why do you have to reset `iter_i`?

Reply:
You can, and I will when I commit in response to your comments. 

Before I factored the code, this was part of a big function where I used a single iterator multiple times to avoid unnecessary allocations. Also, it wasn't clear to me (until today when I closely read the cpp file) that SpherepackIterator's constructor initializes itself, making the reset call unnecessary. 

Finally, note that the reset call sets a single int, so leaving this would not have been a big deal.